sources/cassandra/main.tf

```diff
@@ -0,0 +1,102 @@
+resource "google_compute_firewall" "allow_ports" {
+  name    = "allow-zdm-ports"
+  network = var.common_params.network
+
+  allow {
+    protocol = "tcp"
+    ports    = ["14002", "9042"]
+  }
+
+  source_ranges = ["0.0.0.0/0"]
+  target_tags   = ["zdm"]
+}
+
+resource "google_compute_instance" "cos_instances" {
+  count        = var.common_params.instance_count
+  name         = "${var.common_params.instance_name_prefix}-${count.index}"
+  machine_type = var.common_params.machine_type
+  zone         = var.common_params.zone
+
+  boot_disk {
+    initialize_params {
+      image = "projects/cos-cloud/global/images/cos-105-17412-535-55"
+      size  = 32
+      type  = "pd-balanced"
+    }
+  }
+
+  network_interface {
+    network = var.common_params.network
+    access_config {}
+  }
+
+  tags = ["zdm"]
+
+  metadata = {
+    enable-oslogin = "TRUE"
+  }
+}
+
+resource "local_file" "zdm_config" {
+  count    = var.common_params.instance_count
+  filename = "zdm-config-${count.index}.yaml"
+  content = yamlencode({
+    origin_contact_points = var.zdm_config.origin_contact_points
+    origin_port           = var.zdm_config.origin_port
+    origin_username       = var.zdm_config.origin_username
+    origin_password       = var.zdm_config.origin_password
+
+    target_contact_points = var.zdm_config.target_contact_points
+    target_port           = var.zdm_config.target_port
+    target_username       = var.zdm_config.target_username
+    target_password       = var.zdm_config.target_password
+
+    proxy_listen_address         = var.zdm_config.proxy_listen_address
+    proxy_listen_port            = var.zdm_config.proxy_listen_port
+    proxy_request_timeout_ms     = var.zdm_config.proxy_request_timeout_ms
+    proxy_max_client_connections = var.zdm_config.proxy_max_client_connections
+    proxy_max_stream_ids         = var.zdm_config.proxy_max_stream_ids
+    log_level                    = var.zdm_config.log_level
+    proxy_topology_addresses     = join(",", google_compute_instance.cos_instances[*].network_interface[0].network_ip)
+    proxy_topology_index         = count.index
+  })
+  depends_on = [google_compute_instance.cos_instances]
+}
+
+resource "null_resource" "setup_instances" {
+  count = var.common_params.instance_count
+
+  provisioner "local-exec" {
+    command = <<-EOT
+      for i in {1..3}; do
```

Comment:
Is 3 number of nodes?
If yes, can we make it configurable.

Reply:
This is a retry loop. I faced network timeouts with the scp command when running orchestrating large number of proxies, so this retries scp thrice before failing. Ideally we would want it configurable but did not want to overload params for the user unless we feel we need more.